### ISSUE TITLE
HTTP status client error (422 Unprocessable Entity) for login

### ISSUE THREAD

**Issue by jam1garner** [iliana/eggbug-rs#1]
Hi! Thanks for the great library. Maybe I'm doing something wrong but when I run the `post` example, I get the following error:

```
Error: request error: HTTP status client error (422 Unprocessable Entity) for url (https://cohost.org/api/v1/login)

Caused by:
    HTTP status client error (422 Unprocessable Entity) for url (https://cohost.org/api/v1/login)
```

I've double checked that the provided email/password/page is correct, as well as tried using other pages to authenticate, but I've had no luck. Looking at the login flow via mitmproxy seems to indicate that the login endpoint hasn't changed, so I'm unsure why the 422 error. Still (to me) sounds like the API has had some minor changes, but I can't figure out any as [the login code](https://github.com/iliana/eggbug-rs/blob/cfc28fc860747b0a4b8fdf86615ea11d477490ce/src/client.rs#L65-L90) seems to match what I'm seeing using the website under mitmproxy (I'd love to give screenshots/more info, but unfortunately I don't trust my ability to redact sensitive data from an authentication flow trace due to being 3 am, but if you need more info later I can provide it)

**Comment by iliana** [comment]
A few months ago I unsuccessfully tried to debug an issue with someone where the same code would work for me, but would return “Login Failed” in the 422 error for them. We never figured it out, I think.

I would recommend logging the client_hash value (don’t share it) and seeing if it exactly matches what is generated by the browser for your user. There’s gotta be something subtly wrong here (maybe a base64 encoding problem that I magically am not hitting somehow?).

**Comment by jam1garner** [comment]
Uhhhhh yeah salt is the same (sanity check), but the `clientHash` field doesn't match. don't think it's a base64 encoding error though, going through the data of each there is very little similarity. likely an issue at/before the hashing step.

complete shot in the dark here but the password length for this account is divisible by 16. I can't think of any feature of PBKDF2 that would have that be relevant without some serious misimplementation/misuse... but I'm not seeing any other obvious room for error (maybe the salt is decoded with an incorrect base64 config? wrong number of PBKDF2 iterations? not sure)

willing to help investigate though if you have any theories

**Comment by iliana** [comment]
Last quick check I'll have you try is logging the length of your password. This caught an issue I had where my password in my .env had a `$` in it, and dotenv was doing environment substitutions.

**Comment by jam1garner** [comment]
Logging the password and it matches exactly (can even copy/paste it to log in via the browser). I'm kinda out of guesses. I guess it's slightly weird that password hashing takes like 10 seconds when built with debug mode but I'm not sure that's really surprising to me either

**Comment by iliana** [comment]
Yeah that's normal for PBKDF2. (I should probably configure it to always build that crate in release mode.)

I will start diving into the minified JavaScript mines to see if there's some conditional code that affects how password hashing is done. I don't think there is, though.

**Comment by iliana** [comment]
Does the salt for your email (`https://cohost.org/api/v1/login/salt?email=:email`) contain a hyphen and/or underscore?

**Comment by iliana** [comment]
I am assuming it does. I debugged with a friend who also has a hyphen and/or underscore in their email's salt, and wrote a workaround. The gory details: https://github.com/iliana/eggbug-rs/blob/a5a3ce477a431e8d3f46e2cce943e05e1ece57de/src/client.rs#L112-L126

I just published v0.1.3 with a fix. Let me know if you run into further issues!

**Comment by jam1garner** [comment]
Thank you so much <3 this fixed it
